DOC: add roadmap (#1065)
pysat/pysat: docs/roadmap.rst

.. _roadmap:

Roadmap
=======

The long-term vision for pysat is to make the package suitable for working with
any combination of data. As pysat is intended to support the development of
highly robust and verifiable scientific analysis and processing packages, pysat
must produce each of its features with high quality unit tests and documentation.

This document provides a broad and long term vision of pysat. Specific tasks
associated with this roadmap may be found within the posted
`Issues <https://github.com/pysat/pysat/issues>`_ and
`Projects <https://github.com/pysat/pysat/projects>`_.

An item being on the roadmap does not mean that it will necessarily happen.
During the implementation or testing periods we may discover issues that limit
the feature.

Generality
----------

> pysat's data support is currently focused on space-science data sets, however,

Reviewer: ```suggestion
pysat's data support is currently focused on space-science data sets. However,
```

> pysat's feature set works well on other types of data as well. Where appropriate,

Reviewer: ```suggestion
pysat's feature set also works well on other types of data. Where appropriate,
```

space-science specific features will be generalized for a wider audience.

Data Support
~~~~~~~~~~~~

> The Instrument class currently supports both pandas and xarray formats, covering

Reviewer: ```suggestion
The :py:class:`Instrument` class currently supports both :py:class:`pandas.DataFrame` and :py:class:`xarray.Dataset` formats, covering
```

> one and multi-dimensional data that can be loaded into memory. Even larger data

Reviewer: ```suggestion
1D time-series and multi-dimensional data that can be loaded into memory. Even larger data
```

sets would require that pysat integrate a data format such as Dask. To cover the
needs of any potential user, an ideal solution would be for pysat to implement
a clear public mechanism for users to add their own data formats. Commonalities
observed after integrating dask, pandas, and xarray should provide a viable path
forward for this generalization.

Multiple Data Sources
~~~~~~~~~~~~~~~~~~~~~
The Instrument class is designed to work on a single data source at a time. For
multiple data sources pysat is developing a Constellation class that operates
on multiple Instrument objects and will include methods designed to assist in
merging multiple data sets together. The Constellation class will feature
compatibility with the simpler Instrument object when possible. However, given
the additional complexity when working with multiple sources this may not always
be possible. Long term, we intend on providing functionality that can merge a
Constellation into a 'live' Instrument object for greatest compatibility.

Metapackage
~~~~~~~~~~~
The minimal barriers to entry in open source software allows for a large
variety of packages, each with its own approach to a problem. A disadvantage

> of the setup is the variety of packages may not have been developed with
> interoperability in mind, presenting challenges when attempting to combine

Reviewer: ```suggestion
of the setup is the variety of packages have been developed without
interoperability in mind, presenting challenges when attempting to combine
```

these disparate packages towards a common goal. pysat's versatility when
coupling to data sources may be used to connect these isolated packages
together using pysat an intermediary. Once a package is connected to pysat
then that functionality becomes available to all packages that incorporate
pysat as a source. The value and functionality of this large scale pysat
metapackage increases exponentially with every new connection.

File Support
~~~~~~~~~~~~

> pysat currently supports tracking both data and metadata, as well as the ability

Reviewer: ```suggestion
:py:mod:`pysat` currently supports tracking both data and metadata, as well as the ability
```

Here and elsewhere

to create netCDF4 files, and is capable of maintaining compliance with NASA's

> Space Physics Data Facility (SPDF/ISTP) formatting requirements for NASA

Reviewer: Recommend making this text a link to the formatting requirements.

> Space Physics Data Facility (SPDF/ISTP) formatting requirements for NASA
> satellite missions. Support for creating different types of files, as well as

Author: ```suggestion
`Space Physics Data Facility (SPDF) <https://spdf.gsfc.nasa.gov/sp_use_of_cdf.html>`_
formatting requirements for NASA
satellite missions. Support for creating different types of files, as well as
```

.. _roadmap:

Roadmap
=======

The long-term vision for :py:mod:`pysat` is to make the package suitable for
working with any combination of data. As :py:mod:`pysat` is intended to support
the development of highly robust and verifiable scientific analysis and
processing packages, :py:mod:`pysat` must produce each of its features with high
quality unit tests and documentation.

This document provides a broad and long term vision of :py:mod:`pysat`. Specific
tasks associated with this roadmap may be found within the posted
`Issues <https://github.com/pysat/pysat/issues>`_ and
`Projects <https://github.com/pysat/pysat/projects>`_.

An item being on the roadmap does not necessarily mean that it will happen.
During the implementation or testing periods we may discover issues that limit
the feature.

Data support with :py:mod:`pysat` is currently focused on space-science data
sets. However, the features within the module also work well on other types of
data. Where appropriate, space-science specific features will be generalized for
a wider audience.

> Data Support
> -------------------

Author: ```suggestion
Data Support
------------
```

The :py:class:`Instrument` class currently supports both
:py:class:`pandas.DataFrame` and :py:class:`xarray.Dataset` formats, covering
1D time-series and multi-dimensional data that can be loaded into memory. Even
larger data sets would require that :py:mod:`pysat` integrate a data format such
as Dask. To cover the needs of any potential user, an ideal solution would be
for :py:mod:`pysat` to implement a clear public mechanism for users to add their
own data formats. Commonalities observed after integrating :py:mod:`dask`,
:py:mod:`pandas`, and :py:mod:`xarray` should provide a viable path forward for
this generalization.

Multiple Data Sources

> ~~~~~~~~~~~~~~~~~~~~~

Reviewer: All of these headers need to be "-"

then that functionality becomes available to all packages that incorporate
:py:mod:`pysat` as a source. The value and functionality of this large scale
:py:mod:`pysat` metapackage increases exponentially with every new connection.

File Support
~~~~~~~~~~~~
:py:mod:`pysat` currently supports tracking both data and metadata, as well as
the ability to create netCDF4 files, and is capable of maintaining compliance
with NASA's
`Space Physics Data Facility <https://spdf.gsfc.nasa.gov/sp_use_of_cdf.html>`_
(SPDF) formatting requirements for NASA satellite missions. Support for creating
different types of files, as well as a variety of file standards, needs to be
enhanced to support a broader array of research areas.

Data Iteration
~~~~~~~~~~~~~~
:py:mod:`pysat` currently features orbit iteration, a feature that transparently
provides complete orbits (across day/file breaks) calculated in real time. A
variety of orbit types are supported, each of which maps to a method looking for
a particular signal in the data to trigger upon. However, the current variety of
orbit types is insufficient to address community needs. The underlying class is

> capable of iterating over a wider variety conditions though this type of

Reviewer 2: ```suggestion
capable of iterating over a wider variety of conditions though this type of
```

.. _roadmap:

Roadmap
=======

The long-term vision for :py:mod:`pysat` is to make the package suitable for
working with any combination of data. As :py:mod:`pysat` is intended to support
the development of highly robust and verifiable scientific analysis and
processing packages, :py:mod:`pysat` must produce each of its features with high
quality unit tests and documentation.

This document provides a broad and long term vision of :py:mod:`pysat`. Specific
tasks associated with this roadmap may be found within the posted
`Issues <https://github.com/pysat/pysat/issues>`_ and
`Projects <https://github.com/pysat/pysat/projects>`_.

An item being on the roadmap does not necessarily mean that it will happen.
During the implementation or testing periods we may discover issues that limit
the feature.

> Generality
> ----------

Reviewer: ```suggestion
```

Recommend removing this header.  The word is not used correctly (from what I could see looking up the definition) and the text goes well under the top level header.

Data support with :py:mod:`pysat` is currently focused on space-science data
sets. However, the features within the module also work well on other types of
data. Where appropriate, space-science specific features will be generalized for
a wider audience.

Data Support

> ~~~~~~~~~~~~

Reviewer: ```suggestion
-------------------
```

And below replace all `~` with `-`
